devito/grid.py

```diff
@@ -209,3 +258,118 @@ def __setstate__(self, state):
         for k, v in state.items():
             setattr(self, k, v)
         self._distributor = Distributor(self.shape, self.dimensions)
+
+
+class SubDomain(object):
+
+    """A :class:`Grid` subdomain."""
+
+    name = None
+
+    def __init__(self):
+        if self.name is None:
+            raise ValueError("SubDomain requires a `name`")
+        self._dimensions = None
+
+    def __subdomain_finalize__(self, dimensions, shape):
+        # Create the SubDomain's SubDimensions
+        sub_dimensions = []
+        for k, v in self.define(dimensions).items():
+            if isinstance(v, Dimension):
+                sub_dimensions.append(v)
+            else:
+                try:
+                    # Case ('middle', int, int)
+                    side, thickness_left, thickness_right = v
+                    if side != 'middle':
+                        raise ValueError("Expected side 'middle', not `%s`" % side)
+                    sub_dimensions.append(SubDimension.middle('%si' % k.name, k,
+                                                              thickness_left,
+                                                              thickness_right))
+                except ValueError:
+                    side, thickness = v
+                    if side == 'left':
+                        sub_dimensions.append(SubDimension.left('%sleft' % k.name, k,
+                                                                thickness))
+                    elif side == 'right':
+                        sub_dimensions.append(SubDimension.right('%sright' % k.name, k,
+                                                                 thickness))
+                    else:
+                        raise ValueError("Expected sides 'left|right', not `%s`" % side)
+        self._dimensions = tuple(sub_dimensions)
+
+        # Compute the SubDomain shape
+        self._shape = tuple(s - (sum(d.thickness_map.values()) if d.is_Sub else 0)
+                            for d, s in zip(self._dimensions, shape))
+
+    def __eq__(self, other):
+        if not isinstance(other, SubDomain):
+            return False
+        return self.name == other.name and self.dimensions == other.dimensions
+
+    def __hash__(self):
+        return hash((self.name, self.dimensions))
+
+    def __str__(self):
+        return "SubDomain[%s]" % self.name
```

Comment:
Maybe more explicit to put the subdims in `str` rather than name

Reply:
I don't know what else to attach -- what else would you like to see apart from the name? The dimensions?

Reply:
I would do like 
`"SubDomain %s[%s]"  % (self.name, self.dimensions)`
Like `SubDomain interior[x, y]`